### ISSUE TITLE
How to use generated mask for optical flow dataset

### ISSUE THREAD

**Issue by kaillerC94** [HanLingsgjk/UnifiedGeneralization#3]
Thank you very much for your work. For the optical flow dataset, how can the finally generated mask_DPS and mask_PPF be converted into masks or valid images that can be used in optical flow model training? Is there any relevant code?

**Comment by HanLingsgjk** [comment]
Of course, we plan to release this part with the new project by the end of this month. If you are in a hurry to use it,  I can give you a simple example first.

This is a typical RAFT style data loader, where amask corresponds to the reconstruction confidence mask, smask corresponds to the mechanism similarity mask, occall is the occlusion mask, and pmask is the geometric consistency mask

```
def readNerfMask(filename):
    mask = cv2.imread(filename, cv2.IMREAD_ANYDEPTH|cv2.IMREAD_COLOR)
    mask = mask.astype(np.float32)
    ssim = mask[:, :,0]/16384.0
    d2dloss = mask[:, :,1]/16384.0
    alpha = mask[:, :,2]/16384.0

    return ssim,d2dloss,alpha

def readNerfddc(filename):
    mask = cv2.imread(filename, cv2.IMREAD_ANYDEPTH|cv2.IMREAD_COLOR)
    mask = mask.astype(np.float32)
    d1 = mask[:, :,0]/1024.0
    d2 = mask[:, :,1]/1024.0
    dc = mask[:, :,2]/8192.0

    return d1,d2,dc

def writeFlowKITTI(filename, uv):
    uv = 64.0 * uv + 2**15
    valid = np.ones([uv.shape[0], uv.shape[1], 1])
    uv = np.concatenate([uv, valid], axis=-1).astype(np.uint16)
    cv2.imwrite(filename, uv[..., ::-1])

class GSDataset(data.Dataset):
    def __init__(self, aug_params=None, sparse=True):
        self.augmentor = None
        self.sparse = sparse
        if aug_params is not None:
            if sparse:
                self.augmentor = SparseFlowAugmentorm(**aug_params)
            else:
                self.augmentor = FlowAugmentor(**aug_params)

        self.isnerf = False
        self.flow_list = []
        self.mask_DPS = []
        self.mask_PPF = []
        self.depth_list = []
        self.image_list = []
        self.extra_info = []
        self.GSddc = []
        self.kit = 0
        self.fflow_list =[]
        self.fmask_DPS =[]
        self.fmask_PPF =[]
        self.fGSddc =[]
        self.image_listf = []
        self.fimage12 = []

    def __getitem__(self, index):
        self.kit = self.kit +1
        index = index % len(self.image_list)
        flow, valid = readFlowKITTI(self.flow_list[index])

        pds, rdds, ssim = readNerfMask(self.mask_DPS[index])
        pah1, pah2, occall = readNerfMask(self.mask_PPF[index])

        pmask = np.array(pds < 1).astype(np.float32)
        amask = np.array(pah1 < 0.06).astype(np.float32)
        valid_mask = occall<0.02
        dmask = np.array(rdds < 0.005).astype(np.float32)
        smask = np.array(ssim >0.9).astype(np.float32)

        valid_mask = np.array(valid_mask).astype(np.float32)
        #试一下没有amask的
        valid = amask*dmask*valid_mask

        d1, d2, dc_change = readNerfddc(self.GSddc[index])
        dc_change = np.concatenate((dc_change[:, :, np.newaxis], valid[:, :, np.newaxis]), axis=2)


        img1 = frame_utils.read_gen(self.image_list[index][0])
        img2 = frame_utils.read_gen(self.image_list[index][1])


        flow = np.array(flow).astype(np.float32)
        img1 = np.array(img1).astype(np.uint8)
        img2 = np.array(img2).astype(np.uint8)
        dc_change = np.array(dc_change).astype(np.float32)
        depth1 = np.array(d1).astype(np.float32)
        #如果要纯静态的输出

        if np.random.random(1)<0.01:
            img2  =  img1
            flow = flow*0
            valid = np.ones_like(valid)
            dc_change = np.ones_like(dc_change)


        if self.augmentor is not None:
            img1, img2, flow,dc_change, valid = self.augmentor(img1, img2, flow,dc_change, valid)

        img1 = torch.from_numpy(img1).permute(2, 0, 1).float()
        img2 = torch.from_numpy(img2).permute(2, 0, 1).float()
        flow = torch.from_numpy(flow).permute(2, 0, 1).float()
        dc_change   = torch.from_numpy(dc_change).permute(2, 0, 1).float()
        if valid is not None:
            valid = torch.from_numpy(valid)
        else:
            valid = (flow[0].abs() < 1000) & (flow[1].abs() < 1000)

        return img1, img2, flow, dc_change, valid.float()

    def __rmul__(self, v):
        self.flow_list = v * self.flow_list
        self.image_list = v * self.image_list
        self.mask_DPS = v*self.mask_DPS
        self.mask_PPF = v*self.mask_PPF
        self.depth_list = v * self.depth_list
        self.GSddc = v * self.GSddc
        return self
```

```
class GSdata(GSDataset):
    def __init__(self, aug_params=None, split='training', root='/home/lh/all_datasets/MIPGS10K_flow/',get_depth=0):
        super(GSdata, self).__init__(aug_params, sparse=True)
        self.get_depth=get_depth
        self.occlusion = False
        self.isnerf = True

        for id in range(0,38):
            strid = str(id).zfill(4)
            rootuse = root + strid
            images1 = sorted(glob(osp.join(rootuse, 'image1/*')))
            images2 = sorted(glob(osp.join(rootuse, 'image2/*')))
            for img1, img2 in zip(images1, images2):
                frame_id = img1.split('/')[-1]
                self.extra_info += [[frame_id]]
                self.image_list += [[img1, img2]]

            self.mask_DPS += sorted(glob(osp.join(rootuse, 'mask_DPS/*')))
            self.mask_PPF += sorted(glob(osp.join(rootuse, 'mask_PPF/*')))
            self.GSddc += sorted(glob(osp.join(rootuse, 'depth/*')))
            self.flow_list += sorted(glob(osp.join(rootuse, 'flow/*')))
```
